File: C++/legis.cpp
#include <iostream>
#include <vector>
#include <string>

class Member {
private:
    std::string name, constituency, party, position;
    int age;

public:
    Member(const std::string& name, int age, const std::string& constituency, const std::string& party, const std::string& position)
        : name(name), age(age), constituency(constituency), party(party), position(position) {}

    void display() const {
        std::cout << "Name: " << name << "\nAge: " << age << "\nConstituency: " << constituency
                  << "\nParty: " << party << "\nPosition: " << position << std::endl;
    }

    std::string getConstituency() const {
        return constituency;
    }
};

class Bill {
public:
    enum class Status { PROPOSED, PENDING, APPROVED, REJECTED };

private:
    std::string title, description, proposer, dateIntroduced;
    Status status;

public:
    Bill(const std::string& title, const std::string& description, const std::string& proposer, const std::string& dateIntroduced, Status status)
        : title(title), description(description), proposer(proposer), dateIntroduced(dateIntroduced), status(status) {}

    void display() const {
        std::cout << "Title: " << title << "\nDescription: " << description << "\nProposer: " << proposer
                  << "\nDate Introduced: " << dateIntroduced << "\nStatus: " << static_cast<int>(status) << std::endl;
    }

    static Status stringToStatus(const std::string& s) {
        if (s == "Proposed") return Status::PROPOSED;
        if (s == "Pending") return Status::PENDING;
        if (s == "Approved") return Status::APPROVED;
        return Status::REJECTED;
    }
};

class Session {
private:
    std::string date, agenda;
    int startHour, startMinute, endHour, endMinute;
    std::vector<std::string> topics;

public:
    Session(const std::string& date, int startHour, int startMinute, int endHour, int endMinute, const std::string& agenda)
        : date(date), startHour(startHour), startMinute(startMinute), endHour(endHour), endMinute(endMinute), agenda(agenda) {}

    void addTopic(const std::string& topic) {
        topics.push_back(topic);
    }

    void display() const {
        std::cout << "Date: " << date << "\nStart Time: " << startHour << ":" << startMinute
                  << "\nEnd Time: " << endHour << ":" << endMinute << "\nAgenda: " << agenda << "\nTopics:";
        for (const auto& topic : topics) std::cout << "\n - " << topic;
        std::cout << std::endl;
    }
};

class Constituency {
private:
    std::string name;
    std::vector<Member> members;

public:
    Constituency(const std::string& name) : name(name) {}

    void addMember(const Member& member) {
        members.push_back(member);
    }

    std::string getName() const {
        return name;
    }

    void display() const {
        std::cout << "Constituency: " << name << "\nMembers:";
        for (const auto& member : members) member.display();
        std::cout << std::endl;
    }
};

class Legis {
private:
    std::vector<Member> members;
    std::vector<Bill> bills;
    std::vector<Session> sessions;
    std::vector<Constituency> constituencies;

    template<typename T>
    T readInput(const std::string& prompt) {
        T value;
        std::cout << prompt;
        std::cin >> value;
        std::cin.ignore();
        return value;
    }

    std::string readString(const std::string& prompt) {
        std::string value;
        std::cout << prompt;
        std::getline(std::cin, value);
        return value;
    }

    Bill::Status readBillStatus() {
        std::string statusStr;
        Bill::Status status;
        while (true) {
            statusStr = readString("Enter status (Proposed, Pending, Approved, Rejected): ");
            try {
                status = Bill::stringToStatus(statusStr);
                return status;
            } catch (...) {
                std::cout << "Invalid status. Try again." << std::endl;
            }
        }
    }

    Constituency* findConstituency(const std::string& name) {
        for (auto& c : constituencies) {
            if (c.getName() == name) return &c;
        }
        return nullptr;
    }

public:
    void run() {
        while (true) {
            std::cout << "1. Add Member\n2. View Members\n3. Add Bill\n4. Create Session\n5. View Sessions\n6. View Constituencies\n0. Exit\n";
            int choice = readInput<int>("Enter choice: ");

            switch (choice) {
                case 1: {
                    std::string name = readString("Enter member name: ");
                    int age = readInput<int>("Enter age: ");
                    std::string constituency = readString("Enter constituency: ");
                    std::string party = readString("Enter party: ");
                    std::string position = readString("Enter position: ");
                    members.emplace_back(name, age, constituency, party, position);
                    auto* cons = findConstituency(constituency);
                    if (!cons) {
                        constituencies.emplace_back(constituency);
                        cons = &constituencies.back();
                    }
                    cons->addMember(members.back());
                    break;
                }
                case 2:
                    for (const auto& member : members) member.display();
                    break;
                case 3: {
                    std::string title = readString("Enter bill title: ");
                    std::string description = readString("Enter bill description: ");
                    std::string proposer = readString("Enter proposer: ");
                    std::string dateIntroduced = readString("Enter date introduced: ");
                    auto status = readBillStatus();
                    bills.emplace_back(title, description, proposer, dateIntroduced, status);
                    break;
                }
                case 4: {
                    std::string date = readString("Enter session date: ");
                    int startHour = readInput<int>("Enter start hour: ");
                    int startMinute = readInput<int>("Enter start minute: ");
                    int endHour = readInput<int>("Enter end hour: ");
                    int endMinute = readInput<int>("Enter end minute: ");
                    std::string agenda = readString("Enter agenda: ");
                    Session session(date, startHour, startMinute, endHour, endMinute, agenda);
                    std::string addTopic;
                    do {
                        std::string topic = readString("Enter topic of discussion: ");
                        session.addTopic(topic);
                        addTopic = readString("Add another topic? (yes/no): ");
                    } while (addTopic == "yes");
                    sessions.push_back(session);
                    break;
                }
                case 5:
                    for (const auto& session : sessions) session.display();
                    break;
                case 6:
                    for (const auto& constituency : constituencies) constituency.display();
                    break;
                case 0:
                    return;
                default:
                    std::cout << "Invalid choice." << std::endl;
            }
        }
    }
};

int main() {
    Legis legis;
    legis.run();
    return 0;
}

File: Python/legis.py
#Author:Clayton
#Rollno:31
#Title:Goa Legislative Assembly
#Startdate:15-07-24
#Modifieddate:22-07-24
#Description:to show oop implementation in goa legislative assembly
from enum import Enum

# Member class definition
class Member:
    def __init__(self, name, age, constituency, party, position):
        self.name = name
        self.age = age
        self.constituency = constituency
        self.party = party
        self.position = position

    def display_info(self):
        print(f"Name: {self.name}, Age: {self.age}, Constituency: {self.constituency}, Party: {self.party}, Position: {self.position}")

class BillStatus(Enum):
    PROPOSED = "Proposed"
    PENDING = "Pending"
    APPROVED = "Approved"
    REJECTED = "Rejected"


class Bill:
    def __init__(self, title, description, proposer, date_intro, status):
        self.title = title
        self.description = description
        self.proposer = proposer
        self.date_intro = date_intro
        self.status = status

# Session class definition
class Session:
    def __init__(self, session_date, start_hour, start_minute, end_hour, end_minute, agenda):
        self.session_date = session_date
        self.start_hour = start_hour
        self.start_minute = start_minute
        self.end_hour = end_hour
        self.end_minute = end_minute
        self.agenda = agenda
        self.topics_of_discussion = []

    def add_topic_of_discussion(self, topic):
        self.topics_of_discussion.append(topic)

# Constituency class definition
class Constituency:
    def __init__(self, name):
        self.name = name
        self.members = []

    def add_member(self, member):
        self.members.append(member)

# Main program
def main():
    members = []
    bills = []
    sessions = []
    constituencies = []

    while True:
        print("Welcome to Goa Legislative Assembly")
        print("1. Add Member")
        print("2. View Members")
        print("3. Add Bill")
        print("4. Create Session")
        print("5. See Sessions created")
        print("6. View Constituencies")
        print("0. EXIT")

        choice = int(input("Enter your choice: "))

        if choice == 1:
            add_member(members, constituencies)
        elif choice == 2:
            display_members(members)
        elif choice == 3:
            add_bill(bills, members)
        elif choice == 4:
            create_session(sessions)
        elif choice == 5:
            see_sessions(sessions)
        elif choice == 6:
            view_constituencies(constituencies)
        elif choice == 0:
            print("Exiting program...")
            break
        else:
            print("Invalid choice. Please enter a number from 0 to 6.")

def add_member(members, constituencies):
    name = input("Enter member's name: ")
    age = int(input("Enter member's age: "))
    constituency_name = input("Enter member's constituency: ")
    party = input("Enter member's party: ")
    position = input("Enter member's position: ")

    constituency = create(constituency_name, constituencies)

    member = Member(name, age, constituency, party, position)
    members.append(member)
    constituency.add_member(member)

    print(f"Member {name} successfully added {constituency_name}.")

def create(constituency_name, constituencies):
    for constituency in constituencies:
        if constituency.name == constituency_name:
            return constituency
    # If constituency not found, create a new one
    new_constituency = Constituency(constituency_name)
    constituencies.append(new_constituency)
    return new_constituency

def display_members(members):
    if not members:
        print("No members found.")
    else:
        print("List of Members:")
        for member in members:
            member.display_info()

def add_bill(bills, members):
    title = input("Enter title of the bill: ")
    description = input("Enter description of the bill: ")
    proposer_name = input("Enter name of the bill proposer: ")

    # Find the proposer member object
    proposer = find_member_by_name(proposer_name, members)
    if proposer:
        date_intro = input("Enter date of introduction (YYYY-MM-DD): ")
        status = input("Enter status of the bill (Proposed/Pending/Approved/Rejected): ")

        # Validate status input against BillStatus enum
        status = validate_bill_status(status)

        bill = Bill(title, description, proposer, date_intro, status)
        bills.append(bill)
        print(f"Bill '{title}' added successfully.")
    else:
        print(f"Member '{proposer_name}' not found.")

def find_member_by_name(name, members):
    for member in members:
        if member.name == name:
            return member
    return None

def validate_bill_status(status):
    try:
        return BillStatus[status.upper()].value
    except KeyError:
        print("Invalid bill status. Defaulting to 'Proposed'.")
        return BillStatus.PROPOSED.value

def create_session(sessions):
    session_date = input("Enter session date (YYYY-MM-DD): ")
    start_hour = int(input("Enter starting hour (0-23): "))
    start_minute = int(input("Enter starting minute (0-59): "))
    end_hour = int(input("Enter ending hour (0-23): "))
    end_minute = int(input("Enter ending minute (0-59): "))
    agenda = input("Enter session agenda: ")

    session = Session(session_date, start_hour, start_minute, end_hour, end_minute, agenda)
    sessions.append(session)
    print("Session created successfully.")

def see_sessions(sessions):
    if not sessions:
        print("No sessions created yet.")
    else:
        print("Sessions created:")
        for idx, session in enumerate(sessions, start=1):
            print(f"Session {idx}:")
            print(f"Date: {session.session_date}")
            print(f"Time: {session.start_hour}:{session.start_minute} - {session.end_hour}:{session.end_minute}")
            print(f"Agenda: {session.agenda}")
            if session.topics_of_discussion:
                print("Topics of Discussion:")
                for topic in session.topics_of_discussion:
                    print(f"- {topic}")

def view_constituencies(constituencies):
    if not constituencies:
        print("No constituencies found.")
    else:
        print("List of Constituencies:")
        for constituency in constituencies:
            print(f"- {constituency.name}")

if __name__ == "__main__":
    main()

    
   